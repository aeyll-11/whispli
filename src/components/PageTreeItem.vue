<script lang="ts" setup>
import type { Data } from '@/interfaces/data.interface';
import { computed, ref } from 'vue';

const props = defineProps<{ data: Data }>();

const hasChildren = computed<boolean>(() => Array.isArray(props.data.children));

/**
 * Handle icon display logic
 */
const icon = computed(() => {
  if (hasChildren.value) {
    return '▶';
  }
  return '•';
});

const isOpen = ref<boolean>(false);

const toggle = (): void => {
  if (hasChildren.value) {
    isOpen.value = !isOpen.value;
  }
};
</script>

<template>
  <ul>
    <li @click="toggle" :style="hasChildren ? 'cursor:pointer' : ''">
      <span>
        {{ isOpen ? '▼' : icon }}
      </span>
      {{ props.data.name }}
    </li>
    <template v-if="$props.data.children && isOpen">
      <PageTreeItem v-for="item in $props.data.children" :data="item" :key="item.id" />
    </template>
  </ul>
</template>

<style scoped>
li {
  list-style: none;
}
</style>
