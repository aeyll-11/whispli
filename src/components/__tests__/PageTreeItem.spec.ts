import { mount } from '@vue/test-utils';
import { describe, expect, it } from 'vitest';
import PageTreeItem from '../PageTreeItem.vue';

describe('NavBar', () => {
  it('should display children on click event', async () => {
    const data = {
      id: '3',
      name: 'Office Events',
      children: [
        {
          id: '6',
          name: '2018',
          children: [
            {
              id: '10',
              name: 'Summer Picnic'
            },
            {
              id: '11',
              name: "Valentine's Day Party"
            },
            {
              id: '12',
              name: "New Year's Party"
            }
          ]
        },
        {
          id: '7',
          name: '2017',
          children: [
            {
              id: '13',
              name: 'Company Anniversary Celebration'
            }
          ]
        }
      ]
    };

    const wrapper = mount(PageTreeItem, {
      propsData: {
        data
      }
    });

    await wrapper.find('li').trigger('click');

    expect(wrapper.html()).toMatchSnapshot();
  });
});
