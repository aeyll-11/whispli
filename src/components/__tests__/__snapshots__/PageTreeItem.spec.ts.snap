// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PageTreeItem > should display children on click event 1`] = `
"<ul data-v-0920cad7="">
  <li data-v-0920cad7="" style="cursor: pointer;"><span data-v-0920cad7="">▼</span> Office Events</li>
  <ul data-v-0920cad7="">
    <li data-v-0920cad7="" style="cursor: pointer;"><span data-v-0920cad7="">▶</span> 2018</li>
    <!--v-if-->
  </ul>
  <ul data-v-0920cad7="">
    <li data-v-0920cad7="" style="cursor: pointer;"><span data-v-0920cad7="">▶</span> 2017</li>
    <!--v-if-->
  </ul>
</ul>"
`;
